File: tools/dex_strings.mjs
/**
 * Extract readable strings from Android DEX files (string pool only).
 * Usage: node tools/dex_strings.mjs [modelesDir]
 */
import fs from 'fs';
import path from 'path';

function readUleb128(buf, offset) {
  let result = 0;
  let shift = 0;
  let pos = offset;
  while (true) {
    const b = buf[pos++];
    result |= (b & 0x7f) << shift;
    if ((b & 0x80) === 0) break;
    shift += 7;
  }
  return [result, pos];
}

function parseDexStrings(filePath) {
  const data = fs.readFileSync(filePath);
  if (data.slice(0, 4).toString() !== 'dex\n') {
    throw new Error(`Not a DEX file: ${filePath}`);
  }
  const stringIdsSize = data.readUInt32LE(0x38);
  const stringIdsOff = data.readUInt32LE(0x3c);
  const strings = [];

  for (let i = 0; i < stringIdsSize; i++) {
    const off = data.readUInt32LE(stringIdsOff + i * 4);
    const [, pos] = readUleb128(data, off);
    let end = pos;
    while (data[end] !== 0) end++;
    const s = data.slice(pos, end).toString('utf8');
    if (s) strings.push(s);
  }
  return strings;
}

const dominoRe = /domino|tile|bone|pip|spinner|boneyard|draw|pass|score|round|match|turn|player|board|hand|double|block|fives|capot|trick|ai|bot|computer|solo|multi|flyclops|game/i;

function main() {
  const dexDir = process.argv[2] || path.join(process.cwd(), 'modeles');
  const outDir = path.join(dexDir, 'dex_analysis');
  fs.mkdirSync(outDir, { recursive: true });

  const dexFiles = fs.readdirSync(dexDir).filter((f) => /^classes\d*\.dex$/i.test(f)).sort();
  const all = new Set();

  for (const name of dexFiles) {
    const fp = path.join(dexDir, name);
    console.log(`Parsing ${name} (${Math.round(fs.statSync(fp).size / 1024)} KB)...`);
    const strings = parseDexStrings(fp);
    const unique = [...new Set(strings)];
    unique.forEach((s) => all.add(s));
    fs.writeFileSync(path.join(outDir, `${path.basename(name, '.dex')}_strings.txt`), unique.sort().join('\n'));
    console.log(`  ${unique.length} unique strings`);
  }

  const sorted = [...all].sort();
  const gameStrings = sorted.filter((s) => dominoRe.test(s));
  const classes = sorted.filter((s) => s.startsWith('L') && s.includes(';'));
  const typeNames = sorted.filter((s) => s.includes('/') && !s.startsWith('L') && s.length > 5);

  const report = [
    '# DEX String Analysis Report',
    '',
    `Files: ${dexFiles.join(', ')}`,
    `Total unique strings: ${sorted.length}`,
    '',
    '## Domino / Game related strings',
    '',
    ...gameStrings.map((s) => `- ${s}`),
    '',
    '## Class descriptors (first 400)',
    '',
    ...classes.slice(0, 400).map((s) => `- ${s}`),
    '',
    '## Type-like paths',
    '',
    ...typeNames.slice(0, 200).map((s) => `- ${s}`),
  ].join('\n');

  fs.writeFileSync(path.join(outDir, 'STRING_REPORT.md'), report);
  console.log(`Report: ${path.join(outDir, 'STRING_REPORT.md')}`);
  console.log(`Game-related strings: ${gameStrings.length}`);
  console.log(`Class descriptors: ${classes.length}`);
}

main();

File: tools/dex_strings.py
#!/usr/bin/env python3
"""Extract readable strings and class-like identifiers from Android DEX files."""
import re
import struct
import sys
from pathlib import Path

DEX_MAGIC = b"dex\n"


def read_uleb128(data: bytes, offset: int) -> tuple[int, int]:
    result = 0
    shift = 0
    while True:
        b = data[offset]
        offset += 1
        result |= (b & 0x7F) << shift
        if (b & 0x80) == 0:
            break
        shift += 7
    return result, offset


def parse_dex_strings(path: Path) -> list[str]:
    data = path.read_bytes()
    if not data.startswith(DEX_MAGIC):
        raise ValueError(f"Not a DEX file: {path}")

    string_ids_size = struct.unpack_from("<I", data, 0x38)[0]
    string_ids_off = struct.unpack_from("<I", data, 0x3C)[0]

    strings: list[str] = []
    for i in range(string_ids_size):
        off = struct.unpack_from("<I", data, string_ids_off + i * 4)[0]
        _, pos = read_uleb128(data, off)
        end = data.index(b"\x00", pos)
        try:
            s = data[pos:end].decode("utf-8", errors="replace")
        except Exception:
            continue
        if s:
            strings.append(s)
    return strings


def classify_strings(strings: list[str]) -> dict[str, list[str]]:
    cats = {
        "classes": [],
        "methods": [],
        "domino_keywords": [],
        "game_keywords": [],
        "packages": [],
        "other_interesting": [],
    }
    domino_re = re.compile(r"domino|tile|bone|pip|spinner|boneyard|draw|pass|score|round|match|turn|player|board|hand|double|block|fives|capot|trick|ai|bot|computer|solo|multi", re.I)
    class_re = re.compile(r"^[A-Za-z_$][\w$]*(\/[A-Za-z_$][\w$]*)+$|^[L][\w$/;]+$")

    seen = set()
    for s in strings:
        if s in seen or len(s) < 3:
            continue
        seen.add(s)

        if "/" in s and (";" in s or s.count("/") >= 2):
            cats["classes"].append(s)
        elif domino_re.search(s):
            cats["domino_keywords"].append(s)
        elif s.startswith("L") and ";" in s:
            cats["classes"].append(s)
        elif "." in s and s[0].islower() and len(s.split(".")) >= 3:
            cats["packages"].append(s)

    return cats


def main() -> int:
    dex_dir = Path(sys.argv[1]) if len(sys.argv) > 1 else Path("modeles")
    out_dir = dex_dir / "dex_analysis"
    out_dir.mkdir(exist_ok=True)

    all_strings: list[str] = []
    for dex in sorted(dex_dir.glob("classes*.dex")):
        print(f"Parsing {dex.name} ({dex.stat().st_size // 1024} KB)...")
        strings = parse_dex_strings(dex)
        all_strings.extend(strings)
        (out_dir / f"{dex.stem}_strings.txt").write_text("\n".join(sorted(set(strings))), encoding="utf-8")
        print(f"  {len(strings)} strings")

    unique = sorted(set(all_strings))
    cats = classify_strings(unique)

    report_lines = [
        "# DEX String Analysis Report",
        "",
        f"Total unique strings: {len(unique)}",
        "",
        "## Domino / Game related strings",
        "",
    ]
    for s in sorted(set(cats["domino_keywords"]))[:500]:
        report_lines.append(f"- {s}")

    report_lines += ["", "## Class descriptors (sample)", ""]
    class_samples = [s for s in unique if s.startswith("L") and ";" in s][:300]
    for s in class_samples:
        report_lines.append(f"- {s}")

    report_lines += ["", "## Package-like identifiers", ""]
    for s in sorted(set(cats["packages"]))[:200]:
        report_lines.append(f"- {s}")

    (out_dir / "STRING_REPORT.md").write_text("\n".join(report_lines), encoding="utf-8")
    print(f"Report written to {out_dir / 'STRING_REPORT.md'}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
